### PR TITLE
Registers: support sizes greater than 8 bytes for readSymbolicRegion

### PR DESCRIPTION
**Pull request by humeafo**

### CODE REVIEW

File: src/S2EExecutionStateRegisters.cpp [hunk 1]
```diff
@@ -168,31 +168,36 @@ bool S2EExecutionStateRegisters::readSymbolicRegion(unsigned offset, void *_buf,
...
     // XXX: deal with alignment and overlaps?
-
-    ref<Expr> value = wos->read(offset, size * 8);
-    uint64_t concreteValue;
-    if (!isa<ConstantExpr>(value)) {
-        if (!concretize) {
-            return false;
+    for (unsigned i = 0; i < size; i += 8) {
+        unsigned csize = (size - i) > 8 ? 8 : (size - i);
+        ref<Expr> value = wos->read(offset + i, csize * 8);
```
Comment (vitalych) [inline src/S2EExecutionStateRegisters.cpp]: I think ObjectState::read() can handle sizes > 8 bytes. KLEE has an Expr::Int128 data type.

Comment (humeafo) [reply src/S2EExecutionStateRegisters.cpp]: Yes, ObjectState::read() can handle sizes > 8 bytes, but 
concreteValue = m_concretizer->concretize(value, reason.c_str());
is the problem, concreteValue is uint64_t and the following is being called in m_concretizer->concretize:
toConstant(*this, expression, reason.c_str())->getZExtValue()
will assert when constant size > 8 bytes, so I must rewrite the concretize part or reuse the old codes(the easy way)  just like the PR does.

Comment (vitalych) [reply src/S2EExecutionStateRegisters.cpp]: ok, sounds good. Please explain this in a comment in the code.

**Comment by humeafo** [comment]
squash the two commits to one.

File: src/S2EExecutionStateRegisters.cpp [hunk 2]
```diff
@@ -169,30 +169,40 @@ bool S2EExecutionStateRegisters::readSymbolicRegion(unsigned offset, void *_buf,
...
-    uint64_t concreteValue;
-    if (!isa<ConstantExpr>(value)) {
-        if (!concretize) {
-            return false;
+    // m_concretizer->concretize require value size <= 8, to support data
+    // size > 8, we must do concretization every 8 bytes in a loop, or we
+    // can rewrite m_concretizer->concretize to remove the 8 bytes limitation
+
+    for (unsigned i = 0; i < size; i += 8) {
+        unsigned csize = (size - i) > 8 ? 8 : (size - i);
```
Comment (vitalych) [inline src/S2EExecutionStateRegisters.cpp]: ``sizeof(uint64_t)`` instead of 8 would be better (no magic constants)

**Comment by vitalych** [comment]
Instead of fix1 for branch name, you could have something like read-symb-128 or similar. It would make it easier for those who want to try it out.

**Comment by vitalych** [comment]
Prefixing commit msg with module name is also helpful. E.g.,:
``registers: support sizes greater than 8 bytes for readSymbolicRegion``

**Comment by vitalych** [comment]
Don't forget to run clang-format from the s2e source folder.
``$ CLANG_FORMAT=/home/vitaly/s2e/build/llvm-release/bin/clang-format ./run-clang-format.sh``

**Comment by humeafo** [comment]
@vitalych all fixed as you pointed out except the branch name.

File: src/S2EExecutionStateRegisters.cpp [hunk 3]
```diff
@@ -169,30 +169,40 @@ bool S2EExecutionStateRegisters::readSymbolicRegion(unsigned offset, void *_buf,
...
+        if (!isa<ConstantExpr>(value)) {
+            if (!concretize) {
+                return false;
+            }
+
+            size_t regIndex = offset / sizeof(target_ulong);
+            std::string regName = regIndex < (sizeof(regNames) / sizeof(regNames[0]))
+                                      ? regNames[regIndex]
+                                      : "CPUOffset-" + std::to_string(offset);
+            std::string reason = std::string("access to ") + regName + " register from libcpu helper";
```
Comment (adrianherrera) [inline src/S2EExecutionStateRegisters.cpp]: You can just do “access to “ + regName

I don’t think the explicit std::string constructor is necessary.

Comment (humeafo) [reply src/S2EExecutionStateRegisters.cpp]: Thanks for your review fixed.

**Comment by vitalych** [comment]
@humeafo thanks! Please squash and rerun clang format.

**Comment by humeafo** [comment]
@vitalych done.

**Comment by vitalych** [comment]
@humeafo sorry for the long delay to handle this PR. I've just recently built a testsuite for S2E, so it is a perfect time to check this. I'll do it asap.

**Review by vitalych — APPROVED** [review]

**Pull request merged**